### PR TITLE
Fix crash storing empty pascal string

### LINKED ISSUE #21242: store_pascal_string() crashes when saving an empty string
**Godot version:**
v3.0.5-stable_x11.64, Xubuntu

**Issue description:**
_store_pascal_string()_ crashes the game when trying to save an empty string.

**Steps to reproduce:**
Download the example project.
Run the scene.
The game quits.
Edit the script, write anything in the currently empty string.
Run the game.
The game runs fine.

**Example project:**
[pascal_string_bug.zip](https://github.com/godotengine/godot/files/2303609/pascal_string_bug.zip)

### PR DESCRIPTION
**Pull request by ghost**
Fixes #21242

### CODE REVIEW

File: core/os/file_access.cpp [hunk 1]
```diff
@@ -502,7 +502,8 @@ void FileAccess::store_pascal_string(const String &p_string) {
 
 	CharString cs = p_string.utf8();
 	store_32(cs.length());
-	store_buffer((uint8_t *)&cs[0], cs.length());
+	if (cs.length() > 0)
```
Comment (akien-mga) [inline core/os/file_access.cpp]: Why not `if (p_string.length() == 0) return;` as in `store_string`?

Comment (Noshyaar) [reply core/os/file_access.cpp]: From my understanding, pascal string always has the first 4 bytes set to string length. `get_pascal_string` seems to expect that. That's why I don't use the same fix.

**Comment by reduz** [comment]
I dont get this, it seems more like a bug in store_buffer, as it should not store anything if length() is zero..

File: core/os/file_access.cpp (hunk 1 above)
Comment (dragmz) [reply core/os/file_access.cpp]: I think you don't need the condition if changed to:
````store_buffer((uint8_t *)(const char *)cs, cs.length());````
or
````store_buffer((uint8_t *)cs.get_data(), cs.length());````

**Comment by Noshyaar** [comment]
cc: @hpvb 

I guess it crashed while evaluating the first arg of store_buffer.

I'll close this for now as it's not a direct fix at the cause. According to the meeting, indexing [0] on empty String shouldn't crash. So the actual cause is in the lower level.

```
[22:14:41] <reduz> but store_buffer should do nothing if the argument of length is zero
[22:14:58] <reduz> so it does not make much sense
[22:15:01] <reduz> Noshyaar?
[22:15:16] <dragmz> is cs[0] out of buffer read in case cs.length() == 0?
[22:15:46] <Akien> Yeah was about to ask that
[22:15:49] <reduz> commented
[22:15:57] <reduz> no, that works
[22:16:12] <reduz> but buffer length being zero should result in nothing done
[22:16:42] <Noshyaar> I'll investigate more into it.
[22:16:47] <Akien> I'll check what the stacktrace is
[22:17:53] <Akien> Well it does crash in cowdata
[22:18:02] <Akien> So sounds like out of bounds read
[22:18:54] <Akien> Actually no, the p_string is a null pointer
[22:18:59] <Akien> The issue is further up the backtrace
[22:19:30] <reduz> oh i see
[22:19:40] <reduz> this one happened as a result of TMMs rewrite to string
[22:19:49] <reduz> TMM: around?
[22:19:57] <Akien> reduz: Well it crashes in 3.0.5 too
[22:20:08] <Akien> That's the master backtrace: https://hastebin.com/awucohoxew.shell
[22:20:27] <reduz> Akien: yeah i see why it crashes, but it should work a bit different
[22:20:47] <Akien> 3.0 backtrace: https://hastebin.com/bajaluxivo.shell
[22:22:23] <Akien> So, needs some more investigation. Let's mve to the next PR for now
[...]
[22:52:13] <Noshyaar> Akien: how did you know that p_string was a null pointer?
[22:52:54] <Akien> Noshyaar: with gdb, `frame 2` to inspect the second frame of the backtrace, `print p_string`
[22:53:57] <Akien> It gave me:
[22:53:57] <Akien> $1 = (const String &) @0x7fffffff4548: {_cowdata = {_ptr = 0x0}}
[22:54:28] <Akien> frame 3 actually
[22:54:47] <Akien> (the _File::store_pascal_string one)
[22:55:32] <reduz> the problem is something else thuogh
[22:55:36] <reduz> need to discuss it with tmm
[22:55:50] <reduz> empty string should still allow reading char 0
```

**Pull request closed without merging**